src/util/requireHook.js

```diff
@@ -1,25 +1,87 @@
 'use strict';
 
-var EventEmitter = require('events').EventEmitter;
 var Module = require('module');
+var path = require('path');
 
-var listeners = new EventEmitter();
-
-var origLoad;
+var executedHooks = {};
+var byModuleNameTransformers = {};
+var byFileNamePatternTransformers = [];
+var origLoad = Module._load;
 
 exports.init = function() {
-  origLoad = Module._load;
-  Module._load = function(file) {
-    var moduleExports = origLoad.apply(Module, arguments);
-    listeners.emit(file, moduleExports);
-    return moduleExports;
+  Module._load = patchedModuleLoad;
+};
+
+function patchedModuleLoad(moduleName) {
+  var i;
+
+  // First attempt to always get the module via the original implementation
+  // as this action may fail. The original function populates the module cache.
+  var moduleExports = origLoad.apply(Module, arguments);
+  var filename = Module._resolveFilename.apply(this, arguments);
+
+  // We are not directly manipulating the global module cache because there might be other tools fiddling with
+  // Module._load. We don't want to break any of them.
+  var cacheEntry = executedHooks[filename] = executedHooks[filename] || {
+    moduleExports: moduleExports,
+
+    // We might have already seen and processed, i.e. manipulated, this require statement. This is something we
+    // are checking using these fields.
+    appliedByModuleNameTransformers: [],
+    byFileNamePatternTransformersApplied: false
   };
+
+  var applicableByModuleNameTransformers = byModuleNameTransformers[moduleName];
+  if (applicableByModuleNameTransformers && cacheEntry.appliedByModuleNameTransformers.indexOf(moduleName) === -1) {
+    for (i = 0; i < applicableByModuleNameTransformers.length; i++) {
+      cacheEntry.moduleExports = applicableByModuleNameTransformers[i](cacheEntry.moduleExports) ||
+        cacheEntry.moduleExports;
+    }
+    cacheEntry.appliedByModuleNameTransformers.push(moduleName);
+  }
+
+  if (!cacheEntry.byFileNamePatternTransformersApplied) {
+    for (i = 0; i < byFileNamePatternTransformers.length; i++) {
+      if (byFileNamePatternTransformers[i].pattern.test(filename)) {
+        cacheEntry.moduleExports = byFileNamePatternTransformers[i].fn(cacheEntry.moduleExports) ||
+          cacheEntry.moduleExports;
+      }
+    }
+    cacheEntry.byFileNamePatternTransformersApplied = true;
+  }
+
+  return cacheEntry.moduleExports;
+}
+
+exports.teardownForTestPurposes = function() {
+  Module._load = origLoad;
+  executedHooks = {};
+  byModuleNameTransformers = {};
 };
 
-exports.teardown = function() {
-  Module._load = origLoad || Module._load;
+exports.onModuleLoad = function on(moduleName, fn) {
+  var eventListeners = byModuleNameTransformers[moduleName] = byModuleNameTransformers[moduleName] || [];
+  eventListeners.push(fn);
```

Comment:
We could get rid of the local `eventListeners` var and instead push into `byModuleNameTransformers[moduleName] ` directly, I suppose (makes the two lines a bit easier to grok also, I think).

Reply:
That would mean like this?

```
(byModuleNameTransformers[moduleName] = byModuleNameTransformers[moduleName] || []).push(fn);
```

I don't particularly like this style, I'll gladly change it to this if you prefer?

Reply:
clarified face to face and done.